---
layout: docs
seo_title: About Me
bottom_meta: false
sidebar:
---

# About Me

The About Me page has been rewrited! Now it uses tabs to collapse the content... to save space from being wasted too often~

## Personal Profile
- Self-proclaimed: Silen Wang / Sylens Wong, is a rosted egg that is gradually becoming happier!
- Interests: Playing games, playing games, and still damn playing! Games! Also watch anime, ~~but recently I'm less and less interested in series...~~, In 2024, I started chasing series again! [ダンジョン飯](https://www.bilibili.com/bangumi/media/md21174614) is really amazing! Go watch it!
- Learning/Working Path: Public Health -> Molecular Epidemiology -> Bioinformatics -> Neoantigens -> Organoids -> All kinds of things
  + emmmm... becoming more and more eclectic
  + The word cloud below can't represent me anymore, updating soon...

![Abstract Word Cloud](https://raw.githubusercontent.com/SilenWang/Gallary/master/2026/01/upgit_20260110_1768039212.png)

- GitHub:

<p style="display: flex;">
<img height=180 align="center" src="https://github-readme-stats-ruby-nu-23.vercel.app/api?username=SilenWang&count_private=true&show_icons=true" />
<img height=180 align="center" src="https://github-readme-stats-ruby-nu-23.vercel.app/api/top-langs?username=SilenWang&layout=compact&langs_count=8&card_width=320&hide=html,javascript,css" />
</p>

## Education

{% timeline %}

  {% timenode 2009.09-2014.06 %}

  - School: Guangdong Pharmaceutical University (formerly Guangdong Pharmaceutical College)
  - College: School of Public Health
  - Major: Preventive Medicine
  - Degree: Bachelor

  {% endtimenode %}

  {% timenode 2014.09-2017.06 %}

  - School: Southern Medical University
  - College: School of Public Health
  - Major: Epidemiology and Health Statistics
  - Degree: Master

  {% endtimenode %}

{% endtimeline %}


## Work Experience

{% tabs tab-job %}

<!-- tab 2017.07-2018.08 -->

- Employment: Bioinformatics Analysis Engineer at a bioinformatics service company in Tianjin
- Work Content:
  1. Human genomic next-generation sequencing data (WES / WGS) quality control and analysis
      - Variant result screening based on frequency, gene region, pathogenicity software prediction, and family inheritance patterns
      - Localization of potentially harmful variants based on genome-wide association study (GWAS) results
      - Customization of screening plans and presentation of results according to customer requirements
  2. Research on the current status of specific diseases
      - Research and learn the basic knowledge and current research status of specific diseases to prepare for the development of related analysis modules
  3. Maintenance and upgrade of data delivery processes

<!-- endtab -->

<!-- tab 2018.08-2019.07 -->

- Employment: Bioinformatics Engineer at a genetic technology company in Hangzhou
- Work Content:
  - Data Processing
    1. NGS data processing of cfDNA/FFPE/white blood cell samples
    2. Construction and testing of NGS analysis processes
    3. Development of tools for NGS analysis (Python / R / Julia / Shell) and quality control of cell sample sequencing data
    4. Alignment and quality control of raw sequencing data (fastq)
    5. Processing of UMI data / Molecular Barcode data
  - Analysis Process Related
    1. Debugging of existing bioinformatics analysis processes
    2. Writing new analysis modules and integrating them into the existing pipeline
    3. Testing and debugging of analysis modules
    4. Deployment and dependency management of analysis software
    5. Performance testing of analysis software and comparison of analysis results
    6. Writing and maintaining process documentation
    7. Construction and testing of single-cell sequencing analysis processes
  - Algorithm and Tool Development:
    1. Comparison and replication of UMI deduplication algorithms
    2. Development of tools for drawing quality control plots
    3. Development of variant result screening tools based on ML
    4. Performance optimization of existing tools (parallelization)

<!-- endtab -->

<!-- tab 2019.07-2024.04 -->

- 2019.07-2020.03 Bioinformatics Engineer at a biotechnology company in Hangzhou
  1. Research on NGS-based cfDNA analysis plans, construction of analysis workflows
  2. Statistical analysis of neoantigen-related data
  3. Maintenance of bioinformatics analysis programs
  4. Maintenance of the medication guidance report database
- 2020.03-2024.04 Bioinformatics Supervisor at a biotechnology company in Hangzhou
  1. Development of NGS-based HLA typing/quantification modules
  2. Backend development of bioinformatics analysis tools based on flask
  3. Development of neoantigen screening modules
  4. Maintenance and upgrade of the main bioinformatics analysis workflow
  5. Writing and revising materials as required

<!-- endtab -->

<!-- tab 2024.05-Present -->

- 2024.05-Present Senior Bioinformatics Analysis Engineer at a medical technology company in Hangzhou
  1. Development of WES/RNA-Seq/single‑cell analysis workflows
  2. Research, interpretation, formulation and execution of bioinformatics analysis plans according to customer requirements
  3. Construction, operation and maintenance of bioinformatics databases
  4. Operation and maintenance of bioinformatics servers
  5. Operation and maintenance of websites and system cloud servers
  6. Development, operation, maintenance, migration and automated testing of websites and internal laboratory management systems
  7. Operation and maintenance of WeChat mini‑programs

<!-- endtab -->


{% endtabs %}


## Work-related Skills

### Commonly Used Languages & Tools

#### Programming Languages

{% gallery left %}
<img src="https://img.shields.io/badge/-Python-356f9e?style=flat-square&logo=Python&logoColor=white" />
<img src="https://img.shields.io/badge/-R-276DC3?style=flat-square&logo=R&logoColor=white">
<img src="https://img.shields.io/badge/-Bash-444444?style=flat-square&logo=gnubash&logoColor=white" />
<img src="https://img.shields.io/badge/-Go-007896?style=flat-square&logo=Go&logoColor=white" />
<img src="https://img.shields.io/badge/-HTML-F26524?style=flat-square&logo=html5&logoColor=white" />
<img src="https://img.shields.io/badge/-CSS-633194?style=flat-square&logo=css&logoColor=white" />
<img src="https://img.shields.io/badge/-TypeScript-3178C6?style=flat-square&logo=typescript&logoColor=white" />
{% endgallery %}

#### Data Science / Database

{% gallery left %}
<img src="https://img.shields.io/badge/-Pandas-FFCA00?style=flat-square&logo=pandas&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-Dask-FC6E6B?style=flat-square&logo=dask&logoColor=white" />

<img src="https://img.shields.io/badge/-SQLite-044a64?style=flat-square&logo=sqlite&logoColor=white" />
<img src="https://img.shields.io/badge/-MySQL-F29221?style=flat-square&logo=mysql&logoColor=white" />
<img src="https://img.shields.io/badge/-Neo4j-018bff?style=flat-square&logo=neo4j&logoColor=white" />
<img src="https://img.shields.io/badge/-MongoDB-023430?style=flat-square&logo=mongodb&logoColor=white" />
{% endgallery %}

#### Web
{% gallery left %}
<img src="https://img.shields.io/badge/-Taipy-FF371A?style=flat-square&logo=taipy&logoColor=white" />
<img src="https://img.shields.io/badge/-FastAPI-059487?style=flat-square&logo=fastapi&logoColor=white" />
<img src="https://img.shields.io/badge/-Flask-444444?style=flat-square&logo=flask&logoColor=white" />
<img src="https://img.shields.io/badge/-Jinja-7E0C1B?style=flat-square&logo=jinja&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-Shiny-007BC2?style=flat-square&logo=shinywhite&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-Scrapy-15B8A6?style=flat-square&&logoColor=white&logo=scrapy" />
<img src="https://img.shields.io/badge/-Vue-4FC08D?style=flat-square&logo=vue.js&logoColor=white" />
<img src="https://img.shields.io/badge/-Node-5FA04E?style=flat-square&logo=Node.js&logoColor=white" />
{% endgallery %}

#### IDE / Dev
{% gallery left %}
<img src="https://custom-icon-badges.demolab.com/badge/-VScode-24A4EB?style=flat-square&logo=vscodealt&logoColor=white" />
<img src="https://img.shields.io/badge/-Jupyter-f37524?style=flat-square&logo=Jupyter&logoColor=white" />
<img src="https://img.shields.io/badge/-Rstudio-447099?style=flat-square&logo=rstudioide&logoColor=white" />

<img src="https://custom-icon-badges.demolab.com/badge/-Pixi-ffe300?style=flat-square&logo=prefixdev&logoColor=white" />
<img src="https://img.shields.io/badge/-CondaForge-000000?style=flat-square&logo=condaforge&logoColor=white" />
<img src="https://img.shields.io/badge/-Git-F05032?style=flat-square&logo=git&logoColor=white" />
<img src="https://img.shields.io/badge/-Docker-46a2f1?style=flat-square&logo=docker&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-Devpod-A640FF?style=flat-square&logo=devpod&logoColor=white" />
{% endgallery %}

#### DevOps
{% gallery left %}
<img src="https://img.shields.io/badge/-Github-1F2328?style=flat-square&logo=github&logoColor=white" />
<img src="https://img.shields.io/badge/-Github_Actions-55D263?style=flat-square&logo=githubactions&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-Azure-34B4E8?style=flat-square&logo=azurewhite&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-Azure_Pipelines-34B4E8?style=flat-square&logo=azurepipelinewhite&logoColor=white" />
{% endgallery %}

#### Visualization
{% gallery left %}
<img src="https://img.shields.io/badge/-Plotly-F93778?style=flat-square&logo=plotly&logoColor=white" />
<img src="https://custom-icon-badges.demolab.com/badge/-GGplot2-001F51?style=flat-square&&logoColor=white&logo=ggplot2white" />
{% endgallery %}

#### Bioinformatics
{% gallery left %}
<img src="https://custom-icon-badges.demolab.com/badge/-Snakemake-039475?style=flat-square&&logoColor=white&logo=snakemakewhite" />
<img src="https://custom-icon-badges.demolab.com/badge/-Biopython-3370A3?style=flat-square&&logoColor=white&logo=biopythonwhite" />
<img src="https://custom-icon-badges.demolab.com/badge/-Bedtools-D91E30?style=flat-square&&logoColor=white&logo=bedtoolswhite" />
<img src="https://custom-icon-badges.demolab.com/badge/-GATK-3370A3?style=flat-square&&logoColor=white&logo=gatkwhite" />
<img src="https://custom-icon-badges.demolab.com/badge/-Scanpy-F3643A?style=flat-square&&logoColor=white&logo=scanpywhite" />
{% endgallery %}

#### Operating Systems
{% gallery left %}
<img src="https://custom-icon-badges.demolab.com/badge/-FydeOS-E40046?style=flat-square&logo=fydeos" />
<img src="https://img.shields.io/badge/-Manjaro-35bfa4?style=flat-square&logo=manjaro&logoColor=white" />
<img src="https://img.shields.io/badge/-ChromiumOS-1A73E8?style=flat-square&logo=googlechrome&logoColor=white" />
<img src="https://img.shields.io/badge/-Debian-A80030?style=flat-square&logo=debian&logoColor=white" />
<img src="https://img.shields.io/badge/-Ubuntu-E95420?style=flat-square&logo=ubuntu&logoColor=white" />
{% endgallery %}

### Programming Skills

{% tabs lang %}

<!-- tab Python -->
- Parallelization: Multiprocess
- Next‑generation sequencing file handling: Pysam, PyVCF, CyVCF, Biopython, Scanpy
- Web scraping: 
  + scrapy: Several small‑scale scraping project experiences, capable of using scrapy for static and non‑encrypted dynamic page scraping and data extraction
  + Able to use splash, selenium with scrapy to scrape dynamic page data
- Data processing / statistical computing: Pandas, numpy, math, scipy, statsmodels
- Automated testing: Pytest
- Plotting: Plotly, altair
- Interactive data display: Dash
- Image processing: pillow
- WebUI/API development: FastAPI, Taipy, flask, gradio, nicegui
- Office automation: docxtpl, openpyxl, python‑docx

<!-- endtab -->

<!-- tab R -->

- Data processing: base, dplyr
- Parallelization: Parallel
- Plotting: ggplot2, ggpubr, ggthemes
- Web development: Shiny, plumber

<!-- endtab -->

<!-- tab Workflow Setup -->

- Linux system installation/configuration management: Red‑hat family distribution operation/maintenance, Debian family distribution installation/configuration/maintenance, Arch family distribution installation/configuration/maintenance
- HPC cluster management system: SUN Grid Engine use and maintenance experience
- Snakemake: Proficient in writing Snakefiles, proficient in using Snakemake for workflow construction and management, rich practical Snakemake experience
- Nextflow: Able to understand nextflow configuration files, experience in converting nextflow workflows to Snakemake workflows
- Container technology:
  + Containerized development experience, able to configure development containers compliant with Devcontainer specifications for local or remote development
  + Rich container usage and deployment experience, able to use mainstream container software such as Docker, Podman, Apptainer (formerly Singularity), able to write Dockerfiles and config files to build Docker or Sif containers ([Example](https://github.com/SilenWang/dgidb-docker))
- Pixi/Conda/Mamba: Sufficient practical experience, able to use conda to quickly deploy analysis software and workflows
- Git: Experience using Git for project code/document management, understands basic creation/push/merge methods, has used hook features to run some automated tasks
- DevOps: Understand basic DevOps concepts, able to use Github actions / Azure Pipeline platforms for code management, automated testing and deployment

<!-- endtab -->

{% endtabs %}


### Mastered Bioinformatics Software/Modules

- emmm... I've used quite a few, and I haven't used many of them in the past year... But I should be able to get reacquainted with them all within a week, hahahhahaha

{% tabs bioinfo %}

<!-- tab Data File Processing -->

- Next‑generation sequencing data splitting: bcl2fastq
- BED file manipulation: bedtools
- Data quality control: fastp, fastQC, MultiQC
- Sequence alignment: bwa, bowtie, STAR
- Expression quantification/analysis: DESeq2, DEXseq, Salmon, kallisto, htseq, Hisat2
- Sequence processing: seqtk, seqkit
- BAM/VCF file manipulation: Picard, Samtools, Bcftools, Sambamba, Pysam, Biogo
- UMI processing: UMI‑tools, fgbio

<!-- endtab -->

<!-- tab Single‑Cell Analysis -->

- Analysis frameworks: cellranger, rhapsody, Scanpy, Seurat
- Cellular composition differences: Milopy, miloR
- Cell annotation: SingleR, celltypist, Azimuth
- Pseudotime: Monocle, scvelo, cellrank, pymonocle
- Cell‑cell communication: Cellchat

<!-- endtab -->

<!-- tab Variant Detection & Annotation -->

- SNP/INDEL detection: Samtools, GATK(Mutect2), Vardict, Vacscan, Strelka
- DNA structural variation / fusion gene detection: CREAST, lumpy‑sv, Smoove, SViCT, SvABA, Manta, Delly, GeneFuse, Factera, sv‑tools
- CNV detection: CNVkit, PyLOH
- Variant annotation: ANNOVAR, snpEff, VEP

<!-- endtab -->

<!-- tab Other Analyses -->

- Clonality: PyClone, SciClone, FastClone
- Sequencing data genetic consistency checking: plink, NGSCheckMate
- Mass spectrometry identification: pFind, maxquant

<!-- endtab -->

{% endtabs %}


### Other Work Skills

- Patent draft writing
- Software copyright writing
- Research paper writing
- Static blog/document writing
  - hexo
  - mkdocs
  - gitbook


## Displayable Projects

### Github Projects
| | |
| -- | -- |
|{% ghcard 'SilenWang/FUEX' %}|{% ghcard 'SilenWang/Risk_Region' %}|
|{% ghcard 'SilenWang/CSSN_Spider' %}|{% ghcard 'SilenWang/dgidb-docker' %}|

### Published Articles

I generally only take responsibility for the bioinformatics related parts in the following articles:

1. Ye Feng et al., “[Multi-Epitope Vaccine Design Using an Immunoinformatic Approach for SARS-CoV-2](https://www.ncbi.nlm.nih.gov/pubmed/34208061)” Pathogens 10 (June 11, 2021) 737.
2. Zheling Chen et al., “[A Neoantigen-Based Peptide Vaccine for Patients With Advanced Pancreatic Cancer Refractory to Standard Treatment](https://pubmed.ncbi.nlm.nih.gov/34484187)” Front. Immunol. 12 (2021) 691605.
3. Jiawei Shou et al., “[Combination Treatment of Radiofrequency Ablation and Peptide Neoantigen Vaccination: Promising Modality for Future Cancer Immunotherapy](https://pubmed.ncbi.nlm.nih.gov/36248865)” Front. Immunol. 13 (September 29, 2022) 1000681.
